runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -4,13 +4,32 @@ local ms = require('vim.lsp.protocol').Methods
 local api = vim.api
 local M = {}
 
----@class (private) vim.lsp.inlay_hint.bufstate
+---@class (private) vim.lsp.inlay_hint.globalstate Global state for inlay hints
+---@field enabled boolean Whether inlay hints are enabled for this scope
+---@type vim.lsp.inlay_hint.globalstate
+local globalstate = {
+  enabled = false,
+}
+
+---@class (private) vim.lsp.inlay_hint.bufstate: vim.lsp.inlay_hint.globalstate Buffer local state for inlay hints
 ---@field version? integer
 ---@field client_hints? table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
 ---@field applied table<integer, integer> Last version of hints applied to this line
----@field enabled boolean Whether inlay hints are enabled for this buffer
 ---@type table<integer, vim.lsp.inlay_hint.bufstate>
-local bufstates = {}
+local bufstates = vim.defaulttable(function(_)
```

Comment:
off topic for this PR: 

Usually the correct and strongly preferred way to store buffer-local state is to use a buffer-local variable. Then if a buffer is deleted, its state is also deleted, without needing any lifecycle management. So this `bufstates` pattern should be avoided.

Since you're looking at this code, do you see any reason why these flags can't be stored as a `vim.b` (`b:`) variable? (That change should be a separate PR, if we decide to explore it.)

Reply:
I think it should be possible to store `bufstate` as a `vim.b` variable. Formerly there was no requirement to prepare `bufstate` for every buffer, so we only create them when `enable` inlay hints on a buffer. Maybe it's just because this method is the most direct, and there doesn't seem to be any other LSP-related places using `vim.b`.

Reply:
> Maybe it's just because this method is the most direct, 

Direct how? I would expect that leveraging a buffer-local variable avoids the need for logic to "cleanup" old state in this long-lived container.

> and there doesn't seem to be any other LSP-related places using `vim.b`.

I wouldn't take that as precedent, since (1) it might be driven by the performance consideration mentioned above, and (2) I think LSP features are typically per-client instead of per-buffer.